api/insert/insert.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		return err
+	}
+	r.Body.Close()
+
+	var req jsonAddRequest
+
+	err = json.Unmarshal(body, &req)
+	if err != nil {
+		return errors.NewBadRequestString("Unable to parse certificate addition request")
+	}
+
+	if len(req.Serial) == 0 {
+		return errors.NewBadRequestString("Serial number is required but not provided")
+	}
+
+	if len(req.AKI) == 0 {
+		return errors.NewBadRequestString("Authority key identifier is required but not provided")
+	}
+
+	if _, present := ocsp.StatusCode[req.Status]; !present {
+		return errors.NewBadRequestString("Invalid certificate status")
+	}
+
+	if ocsp.StatusCode[req.Status] == stdocsp.Revoked && req.RevokedAt == (time.Time{}) {
+		return errors.NewBadRequestString("Revoked certificate should specify when it was revoked")
+	}
+
+	if _, present := validReasons[req.Reason]; !present {
+		return errors.NewBadRequestString("Invalid certificate status reason code")
+	}
+
+	if len(req.PEM) == 0 {
+		return errors.NewBadRequestString("The provided certificate is empty")
+	}
+
+	// Parse the certificate and validate that it matches
+	cert, err := helpers.ParseCertificatePEM([]byte(req.PEM))
+	if err != nil {
+		return errors.NewBadRequestString("Unable to parse PEM encoded certificates")
+	}
+
+	serialBigInt := new(big.Int)
+	if _, success := serialBigInt.SetString(req.Serial, 16); !success {
+		return errors.NewBadRequestString("Unable to parse serial key of request")
+	}
+
+	if serialBigInt.Cmp(cert.SerialNumber) != 0 {
+		return errors.NewBadRequestString("Serial key of request and certificate do not match")
+	}
+
+	aki, err := hex.DecodeString(req.AKI)
+	if err != nil {
+		return errors.NewBadRequestString("Unable to decode authority key identifier")
+	}
+
+	if !bytes.Equal(aki, cert.AuthorityKeyId) {
+		return errors.NewBadRequestString("Authority key identifier of request and certificate do not match")
+	}
+
+	cr := certdb.CertificateRecord{
+		Serial:    req.Serial,
+		AKI:       req.AKI,
+		CALabel:   req.CALabel,
+		Status:    req.Status,
+		Reason:    req.Reason,
+		Expiry:    req.Expiry,
+		RevokedAt: req.RevokedAt,
+		PEM:       req.PEM,
+	}
+
+	err = h.dbAccessor.InsertCertificate(cr)
+	if err != nil {
+		return err
+	}
+
+	if h.signer != nil {
+		// Now create an appropriate OCSP response
+		sr := ocsp.SignRequest{
+			Certificate: cert,
+			Status:      req.Status,
+			Reason:      req.Reason,
+			RevokedAt:   req.RevokedAt,
+		}
+		ocspResponse, err := h.signer.Sign(sr)
+		if err != nil {
+			return err
+		}
+
+		// We parse the OCSP repsonse in order to get the next
+		// update time/expiry time
+		ocspParsed, err := stdocsp.ParseResponse(ocspResponse, nil)
+		if err != nil {
+			return err
+		}
+
+		ocspRecord := certdb.OCSPRecord{
+			Serial: req.Serial,
+			AKI:    req.AKI,
+			Body:   string(ocspResponse),
+			Expiry: ocspParsed.NextUpdate,
+		}
+
+		if err = h.dbAccessor.InsertOCSP(ocspRecord); err != nil {
+			return err
+		}
+	}
+
+	result := map[string]string{}
```

Comment:
Oh whoops, I mixed up OCSP and CT stapling.

For `"success": true`, doesn't the fact that we use `api.SendResponse` already include that? This is how a response currently looks:
```
{"success":true,"result":{"ocsp_response":"MIIE2goBAKCCBNMwggTPBgkrBgEFBQcwAQEEggTAMIIEvDCBr6EmMCQxIjAgBgNVBAoTGUNvcm5lbGwgQ1MgNTE1MiBSZXNwb25kZXIYDzIwMTcwMzAyMTUxMTAwWjB0MHIwSjAJBgUrDgMCGgUABBQjMeG40BC+y9MwFMu1wlMuRbbG6QQUyTAQC0ffA+MPxMYWqutvnjIcEP4CEQD/HB5oBaqHa5iDYeQzzNNxgAAYDzIwMTcwMzAyMTUwMDAwWqARGA8yMDE3MDMwMjE2MDAwMFowDQYJKoZIhvcNAQELBQADggEBAFKTiCwk2hv4XmicNgCN5OzHuzL6ETdxH/L6alnPia+tEM13HwHnNNQ9l0iOtcS1ynvxQJizGnWip7Yy29/1Yk3OT2NTqp8svp7pk7n38Y1PXoRrXGF/GWhPcR/AHvneUmc0+HMzvHCSR5njT/hoXFZLBKHvWPiEKMjd86Jc+dQ9MaWYoChkShfl193CVH9VU2jrDDhnXk2G/CbjfRjKdSXrMDUheoCYd7M6d8aO1Yq1TB3HPnOjf1wrDRwXJNYdLHJx9y75vc8P4vK2UTVGd3pVKJ6at+oQvaecpy82p9Nufh98dZNnjLr0M+NmFlN7jJthvl3w4NpRBeZGBs9lfa+gggLyMIIC7jCCAuowggHSoAMCAQICEQDEfqDVUUMWfUMHx4oKsDMoMA0GCSqGSIb3DQEBCwUAMCQxIjAgBgNVBAoTGUNvcm5lbGwgQ1MgNTE1MiBSZXNwb25kZXIwIhgPMDAwMTAxMDEwMDAwMDBaGA8wMDAxMDEwMTAwMDAwMFowJDEiMCAGA1UEChMZQ29ybmVsbCBDUyA1MTUyIFJlc3BvbmRlcjCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAJVDcQNCkhdNCSPmcOwfbpalc7VHma1NViAVcG1OrjPuLVpeYnll7pM3VAZ/0vpT0bPSuPmMSzJH9NG5EOHqzBoMve0/Babk7rk7X+qgSW5ROSAHKoyCFR0IQL1ELELzGmyzUUzbAObZqBPCMQW8ceLqD42rFuLJlCCIvxr/4+LmcC3io7If9V34l6gI3FUfQ3uWl8At3LO0gzBJ/PQ7OktZkZPbDz69moZXSFs4nbAkI39EtivI7LLmFHt3qqr+Su7bfoVboD96iQfuc5bR8MJj3Ses8SR+CQSWxHd/tY/LsGDA0O/gAo7L7BR+0V/pDBOfuWMn8ZpNpGaOoPz4TG8CAwEAAaMTMBEwDwYDVR0jBAgwBoAEKioqKzANBgkqhkiG9w0BAQsFAAOCAQEAfdzDT6yQA9I+6bIM/wKk3BOfJ6PTefrek4iamW56+fy3/bD66Qkl/18xoGrd/MspQy1b1TXOrNQiDYMY62bDDRjBBR5m0IP1N6OT2p30E3APE0KU0obqUXhFcfr1XjQUR+ImDNoIFUMXxVQ4sI3M5emSDNzq0aZhsAyiDwzdaf2Rd50Qzio9h3p3rqgtQBGQ/j5upXRT+HTXNsVq5hFXbycUcZJCDjSkaq0Z7G2OJclrpCdmd0mxFWI7yBIoDncL9nnohsiyGXV2F2RyrtXVaW4JzZAeklP/gAfn0ROdihDd6OIBtYIb4KsTy/uXrvoC6R4b9XF539lKh0Y1+pW9dA=="},"errors":[],"messages":[]}
```

Quick question - in the case that the certificate being added is not signed by the issuer, we can't create an OCSP response. Should this be considered an error in our response to the client? Or should we use something like `api.SendResponseWithMessage` to indicate that?